internal/commands/build.go

```diff
@@ -0,0 +1,259 @@
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/docker/buildx/driver"
+
+	"github.com/docker/buildx/util/progress"
+
+	cnab "github.com/deislabs/cnab-go/driver"
+	"github.com/docker/buildx/bake"
+	"github.com/docker/buildx/build"
+	_ "github.com/docker/buildx/driver/docker" // required to get default driver registered, see driver/docker/factory.go:14
+	"github.com/docker/cli/cli"
+	"github.com/docker/cli/cli/command"
+	dockerclient "github.com/docker/docker/client"
+	"github.com/moby/buildkit/util/appcontext"
+	"github.com/spf13/cobra"
+)
+
+type buildOptions struct {
+	noCache  bool
+	progress string
+	pull     bool
+	tag      string
+	out		 string
+}
+
+func buildCmd(dockerCli command.Cli) *cobra.Command {
+	var opts buildOptions
+	cmd := &cobra.Command{
+		Use:     "build [APPLICATION]",
+		Short:   "Build service images for the application",
+		Example: `$ docker app build myapp.dockerapp`,
+		Args:    cli.RequiresRangeArgs(1, 1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			tag, err := runBuild(dockerCli, args[0], opts)
+			if err == nil {
+				fmt.Printf("Successfully build %s\n", tag.String())
+			}
+			return err
+		},
+	}
+
+	flags := cmd.Flags()
+	flags.BoolVar(&opts.noCache, "no-cache", false, "Do not use cache when building the image")
+	flags.StringVar(&opts.progress, "progress", "auto", "Set type of progress output (auto, plain, tty). Use plain to show container output")
+	flags.BoolVar(&opts.pull, "pull", false, "Always attempt to pull a newer version of the image")
+	flags.StringVarP(&opts.out, "output", "o", "", "Dump generated bundle into a file")
+	flags.StringVarP(&opts.tag, "tag", "t", "", "Name and optionally a tag in the 'name:tag' format")
+
+	return cmd
+}
+
+func runBuild(dockerCli command.Cli, application string, opt buildOptions) (reference.Named, error) {
+	app, err := packager.Extract(application)
+	if err != nil {
+		return nil, err
+	}
+	defer app.Cleanup()
+	appname := app.Name
+
+	bundle, err := makeBundleFromApp(dockerCli, app, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := appcontext.Context()
+
+	compose, err := bake.ParseCompose(app.Composes()[0]) // Fixme can have > 1 composes ?
+	if err != nil {
+		return nil, err
+	}
+
+	targets := map[string]bake.Target{}
+	for _, n := range compose.ResolveGroup("default") {
+		t, err := compose.ResolveTarget(n)
+		if err != nil {
+			return nil, err
+		}
+		if t != nil {
+			targets[n] = *t
+		}
+	}
+
+	for service, t := range targets {
+		if strings.HasPrefix(*t.Context, ".") {
+			// Relative path in compose file under x.dockerapp refers to parent folder
+			// FIXME docker app init should maybe udate them ?
+			path, err := filepath.Abs(appname + "/../" + (*t.Context)[1:])
+			if err != nil {
+				return nil, err
+			}
+			t.Context = &path
+			targets[service] = t
+		}
+	}
+
+	if logrus.IsLevelEnabled(logrus.DebugLevel)	{
+		dt, err := json.MarshalIndent(targets, "", "   ")
+		if err != nil {
+			return nil, err
```

Comment:
Does it mean that if debug is enabled we can fail where it will work if not in debug?
I see two options to have the same path in debug and not:
- run the same code and only print is different
- do not fail but trace the error ("could not marshal targets") and continue

Reply:
true. Anyway if we can't marshal those there's few chance we can run much further.